### PR TITLE
feat(eslint-plugin): allow using void as a default type for a generic argument if allowInGenericTypeArguments is specified

### PR DESCRIPTION
**Pull request by kazizi55**
## PR Checklist

- [x] Addresses an existing open issue: fixes https://github.com/typescript-eslint/typescript-eslint/issues/5644
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

- allows using void as a default type for a generic argument if allowInGenericTypeArguments is specified

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts [hunk 1]
```diff
@@ -162,7 +154,7 @@ ruleTester.run('allowInGenericTypeArguments: true', rule, {
         {
           messageId: 'invalidVoidNotReturnOrGeneric',
           line: 1,
-          column: 41,
+          column: 34,
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts:157]: We still have two errors now, but they're in the same place. That seems wrong.

https://deploy-preview-5671--typescript-eslint.netlify.app/play#ts=4.8.3&sourceType=module&code=MYewdgzgLgBAhgJwSA7gcQKZgwglsAJhgF4YAeAFRgwA8osATCGANxFwZNfYYD4AKRAHMAXDAoBKErxgBvAFAwYoSCAA2GAHRqQQwQiESA3PIC+RoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KImjQO0SODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
The code looks good at first glance, but there's an odd bug in the edge cases. I can take a closer look once that's fixed. Thanks!

File: packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts (hunk 1 above)
Comment (kazizi55) [reply packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts]: @JoshuaKGoldberg 
I fixed the bug and test!
Please check the below! 🙏

https://github.com/typescript-eslint/typescript-eslint/pull/5671/commits/5225063e5a5a13a6a831c49b16a43bf703bdee4a

https://github.com/typescript-eslint/typescript-eslint/pull/5671/commits/2ed96e1f5f84d749b1dfcf3b11a85abd9f0dded1

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts [hunk 2]
```diff
@@ -127,6 +127,35 @@ export default util.createRule<[Options], MessageIds>({
...
 
+    /**
+     * @brief checks if the generic type parameter extends void
+     */
+    function checkExtendedVoid(
+      node: TSESTree.TSVoidKeyword,
+      parentNode: TSESTree.TSTypeParameter,
+    ): void {
+      if (
+        parentNode.constraint?.type !== AST_NODE_TYPES.TSVoidKeyword ||
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-invalid-void-type.ts:138]: I'm confused, why do we care about `parentNode.constraint`? Shouldn't we exclusively care about `parentNode.default`?

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts [hunk 3]
```diff
@@ -127,6 +127,35 @@ export default util.createRule<[Options], MessageIds>({
...
+        JSON.stringify(parentNode.default?.range) !== JSON.stringify(node.range)
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-invalid-void-type.ts:139]: It seems like this is trying to check whether `parentNode.default` _is_ `node`. You can do that directly:

```suggestion
       parentNode.default?.range === node.range
```

Am I missing something?

If I'm understanding right, then the function body could look like:

```ts
if (parentNode.default !== node) {
  context.report({
    messageId: 'invalidVoidNotReturnOrGeneric',
    node,
  });
}
```

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts [hunk 4]
```diff
@@ -127,6 +127,35 @@ export default util.createRule<[Options], MessageIds>({
...
+    }
+
+    /**
+     * @brief checks if the type of given argument is TSESTree.TSTypeParameter
+     */
+    function isTSTypeParameter(
+      node: TSESTree.Node,
+    ): node is TSESTree.TSTypeParameter {
+      return node.type === AST_NODE_TYPES.TSTypeParameter;
+    }
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-invalid-void-type.ts:157]: Nit: this `isTSTypeParameter` function is only used in one place. There's no need to make an extra function for it. You can inline `node.parent.type === AST_NODE_TYPES.TSTypeParameter` where this would be called, below.

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts [hunk 5]
```diff
@@ -127,6 +127,35 @@ export default util.createRule<[Options], MessageIds>({
       }
     }
 
+    /**
+     * @brief checks if the generic type parameter extends void
+     */
+    function checkExtendedVoid(
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-invalid-void-type.ts:133]: Nit: this isn't checking _extension_, no?

```suggestion
    /**
     * @brief checks if the generic type parameter defaults to void
     */
    function checkDefaultVoid(
```

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Getting closer, thanks! The tests all look good but the source code confused me a bit. I think we can simplify it to be more clear.

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts (hunk 3 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts]: Oops, just realized I left the `.range`s in there. Sorry! I'd meant to suggest:

```ts
parentNode.default === node
```

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts (hunk 5 above)
Comment (kazizi55) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts]: You're right!
I fixed!
https://github.com/typescript-eslint/typescript-eslint/pull/5671/commits/a1421824ec0864690a1e6ebf3750e0e75e39096d

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts (hunk 4 above)
Comment (kazizi55) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts:143]: Thank you!
I deleted the function and inlined it, so please check the below:
https://github.com/typescript-eslint/typescript-eslint/pull/5671/commits/4b7ad4127dae4c6f05623ae86eac77c157bcee36

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts (hunk 3 above)
Comment (kazizi55) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts]: Thank you for suggesting!
You're right, so I rewrite the function body as you suggested:
https://github.com/typescript-eslint/typescript-eslint/pull/5671/commits/be5c6f406011fc55a7d4793059099e358be4104c

https://github.com/typescript-eslint/typescript-eslint/pull/5671/commits/508ab8f52d939b10480b97a6cc1f47f85bd95768

File: packages/eslint-plugin/src/rules/no-invalid-void-type.ts (hunk 2 above)
Comment (kazizi55) [reply packages/eslint-plugin/src/rules/no-invalid-void-type.ts]: Sorry to confuse you, that's because of my lack of understanding...
I deleted `parentNode.constraint` from if statement.

**Review by JoshuaKGoldberg — APPROVED** [review]
Super, thanks for sending this in & iterating @kazizi55! 🙌
